src/azure-cli/azure/cli/command_modules/util/custom.py

```diff
@@ -27,3 +29,120 @@ def show_version(cmd):  # pylint: disable=unused-argument
+                exit_code = subprocess.call(az_update_cmd)
+        elif installer == 'RPM':
+            from azure.cli.core.util import get_linux_distro
+            distname, _ = get_linux_distro()
+            if not distname:
+                logger.warning(UPGRADE_MSG)
+            else:
+                distname = distname.lower().strip()
+                if any(x in distname for x in ['centos', 'rhel', 'red hat', 'fedora']):
+                    update_cmd = 'yum update -y azure-cli'.split()
+                    if os.geteuid() != 0:  # pylint: disable=no-member
+                        update_cmd.insert(0, 'sudo')
+                    logger.debug("Update azure cli with '%s'", " ".join(update_cmd))
+                    exit_code = subprocess.call(update_cmd)
+                elif any(x in distname for x in ['opensuse', 'suse', 'sles']):
+                    zypper_refresh_cmd = ['zypper', 'refresh']
+                    az_update_cmd = 'zypper update -y azure-cli'.split()
+                    if os.geteuid() != 0:  # pylint: disable=no-member
+                        zypper_refresh_cmd.insert(0, 'sudo')
+                        az_update_cmd.insert(0, 'sudo')
+                    exit_code = subprocess.call(zypper_refresh_cmd)
+                    if exit_code == 0:
+                        logger.debug("Update azure cli with '%s'", " ".join(az_update_cmd))
+                        exit_code = subprocess.call(az_update_cmd)
+                else:
+                    logger.warning(UPGRADE_MSG)
+        elif installer == 'HOMEBREW':
+            logger.warning("Update homebrew formulae")
+            exit_code = subprocess.call(['brew', 'update'])
+            if exit_code == 0:
+                update_cmd = ['brew', 'upgrade', 'azure-cli']
+                logger.debug("Update azure cli with '%s'", " ".join(update_cmd))
+                exit_code = subprocess.call(update_cmd)
+        elif installer == 'PIP':
+            pip_args = [sys.executable, '-m', 'pip', 'install', '--upgrade', 'azure-cli', '-vv',
+                        '--disable-pip-version-check', '--no-cache-dir']
+            logger.debug("Update azure cli with '%s'", " ".join(pip_args))
+            exit_code = subprocess.call(pip_args, shell=platform.system() == 'Windows')
+        elif installer == 'DOCKER':
+            logger.warning("Exit the container to pull latest image with 'docker pull mcr.microsoft.com/azure-cli' "
+                           "or run 'pip install --upgrade azure-cli' in this container")
+        elif installer == 'MSI':
+            logger.debug("Update azure cli with MSI from https://aka.ms/installazurecliwindows")
+            exit_code = subprocess.call(['powershell.exe', "Start-Process msiexec.exe -Wait -ArgumentList '/i https://aka.ms/installazurecliwindows'"])  # pylint: disable=line-too-long
+        else:
+            logger.warning(UPGRADE_MSG)
+    if exit_code:
+        telemetry.set_failure("CLI upgrade failed.")
```

Comment:
Could we consider record the error information when the upgrade fails in the telemetry?

Reply:
Good suggestion. There is an error improvement PR that revamps telemetry. I will work with @houk-ms to improve error tracking here.